### PR TITLE
feat: add support for Swift macros

### LINKED ISSUE #700: feat: Add support for packages with Swift macros
Swift macros were introduced in Swift 5.9: https://docs.swift.org/swift-book/documentation/the-swift-programming-language/macros/

rules_swift added support for Swift macros in 1.11.0: https://github.com/bazelbuild/rules_swift/blob/master/doc/rules.md#swift_compiler_plugin

Swift packages with macros are gaining popularity and adoption, with a collection of macro packages available here, which could be good candidates for examples and integration tests: https://github.com/krzysztofzablocki/Swift-Macros

### PR DESCRIPTION
**Pull request by jpsim**
* Add support for packages using Swift macros (fixes #700)
* Update `swift-composable-architecture` to 1.4.0 in the TCA example

### CODE REVIEW

File: examples/tca_example/.bazelignore [hunk 1]
```diff
@@ -0,0 +1 @@
+.build
```
Comment (jpsim) [inline examples/tca_example/.bazelignore:1]: swift-syntax defines BUILD files so this is needed to prevent `do_test` from finding it and failing with this error:

```
ERROR: error loading package under directory '': error loading package '.build/checkouts/swift-syntax': Label '//utils/bazel:swift_syntax_library.bzl' is invalid because 'utils/bazel' is not a package; perhaps you meant to put the colon here: '//:utils/bazel/swift_syntax_library.bzl'?
```

File: swiftpkg/internal/swiftpkg_build_files.bzl [hunk 2]
```diff
@@ -51,6 +51,16 @@ def _swift_target_build_file(pkg_ctx, target):
...
     }
+
+    # Add macros as plugins
+    macro_target_labels = [target.label.repository_name + "//:" + target.label.name for target in pkg_ctx.pkg_info.targets if target.type == "macro"]
+    if macro_target_labels:
+        plugins = [target_label for target_label in macro_target_labels for dep in deps if target_label in dep.value[0]]
+        if plugins:
+            attrs["plugins"] = plugins
+            deps_without_plugins = [dep for dep in deps if dep.value[0] not in plugins]
+            attrs["deps"] = bzl_selects.to_starlark(deps_without_plugins)
```
Comment (jpsim) [inline swiftpkg/internal/swiftpkg_build_files.bzl:62]: Suggestions for how to refactor this are appreciated.

File: shared.bazelrc [hunk 3]
```diff
@@ -15,6 +15,9 @@ build --incompatible_strict_action_env=true
 common --enable_bzlmod
 build --@cgrindel_bazel_starlib//bzlmod:enabled
 
+# Set minimum macOS version
+build --macos_minimum_os=13 --host_macos_minimum_os=13
```
Comment (jpsim) [inline shared.bazelrc:19]: If CI passes with this, I'll see if I can remove it or at least lower the minimum version.

Comment (jpsim) [reply shared.bazelrc]: I was able to lower the minimum version to 10.15 and set this in `tca_example`'s `.bazelrc`.

File: examples/tca_example/.bazelrc [hunk 4]
```diff
@@ -6,3 +6,6 @@ import %workspace%/../../ci.bazelrc
 
 # Try to import a local.rc file; typically, written by CI
 try-import %workspace%/../../local.bazelrc
+
+# Set minimum macOS version
+build --macos_minimum_os=10.15 --host_macos_minimum_os=10.15
```
Comment (cgrindel) [inline examples/tca_example/.bazelrc:11]: Curious. Why is this needed?

File: swiftpkg/internal/swiftpkg_build_files.bzl [hunk 5]
```diff
@@ -51,6 +51,16 @@ def _swift_target_build_file(pkg_ctx, target):
         "srcs": pkginfo_targets.srcs(target),
         "visibility": ["//visibility:public"],
     }
+
+    # Add macros as plugins
+    macro_target_labels = [target.label.repository_name + "//:" + target.label.name for target in pkg_ctx.pkg_info.targets if target.type == "macro"]
```
Comment (cgrindel) [inline swiftpkg/internal/swiftpkg_build_files.bzl:56]: Could we wrap this to stay [under 80 chars](https://bazel.build/rules/bzl-style#line-length)?

File: swiftpkg/internal/swiftpkg_build_files.bzl [hunk 6]
```diff
@@ -51,6 +51,16 @@ def _swift_target_build_file(pkg_ctx, target):
...
+    if macro_target_labels:
+        plugins = [target_label for target_label in macro_target_labels for dep in deps if target_label in dep.value[0]]
```
Comment (cgrindel) [inline swiftpkg/internal/swiftpkg_build_files.bzl:58]: [Line length](https://bazel.build/rules/bzl-style#line-length)

File: swiftpkg/internal/swiftpkg_build_files.bzl (hunk 2 above)
Comment (cgrindel) [reply swiftpkg/internal/swiftpkg_build_files.bzl:62]: It looks reasonable to me. Unfortunately, the API for assembling Bazel select statements is a little convoluted due to how they need to be gathered and merged.

**Review by cgrindel — COMMENTED** [review]
Overall, LGTM. My comments are formatting related and a question. Also, I just want to confirm that TCA does use macros. If we did not add macro support, would the TCA upgrade fail?

**Comment by jpsim** [comment]
> Also, I just want to confirm that TCA does use macros. If we did not add macro support, would the TCA upgrade fail?

Yes. `@Reducer` is a new macro in TCA 1.4.0 and App.swift in our example uses it.

File: examples/tca_example/.bazelrc (hunk 4 above)
Comment (jpsim) [reply examples/tca_example/.bazelrc:11]: Fails with this error without this line:

```
dyld[91205]: Symbol not found: _$sSa12_endMutationyyFyXl_Ts5
  Referenced from: <217B6508-EB95-3AF5-9BD0-E76EB0CE1E2B> /private/var/tmp/_bazel_jp/f0f5d2b7b41704886b6a88004e7b18f1/execroot/_main/bazel-out/darwin_arm64-opt-exec-2B5CBBC6-ST-5a29d6cb7794/bin/external/rules_swift_package_manager~override~swift_deps~swiftpkg_swift_case_paths/Sources_CasePathsMacros (built for macOS 14.0 which is newer than running OS)
  Expected in:     <EAA43988-4F04-3D39-A0A6-D597E46E966E> /usr/lib/swift/libswiftCore.dylib
```

File: swiftpkg/internal/swiftpkg_build_files.bzl (hunk 5 above)
Comment (jpsim) [reply swiftpkg/internal/swiftpkg_build_files.bzl]: 037aa33

File: swiftpkg/internal/swiftpkg_build_files.bzl (hunk 6 above)
Comment (jpsim) [reply swiftpkg/internal/swiftpkg_build_files.bzl]: 037aa33

**Comment by jpsim** [comment]
I’m having trouble understanding the CI failures and could use some help. `do_test` in the TCA example directory passes locally for me.

**Comment by jpsim** [comment]
Is it possible that `${{  secrets.BUILDBUDDY_API_KEY }}` isn't resolving for this PR because it's coming from my fork?

https://github.com/cgrindel/rules_swift_package_manager/blob/6023ab6055d225c64cff50f73da87dd78d3491d3/.github/workflows/ci.yml#L139

**Comment by cgrindel** [comment]
> Is it possible that ${{  secrets.BUILDBUDDY_API_KEY }} isn't resolving for this PR because it's coming from my fork?

Possible. However, I did not think that is how GH workflows worked. If you want to remove that from the equation, comment out the remote cache flags in the `ci.bazelrc` file.

**Comment by jpsim** [comment]
> Due to the dangers inherent to automatic processing of PRs, GitHub’s standard `pull_request` workflow trigger by default prevents write permissions and secrets access to the target repository. However, in some scenarios such access is needed to properly process the PR. To this end the `pull_request_target` workflow trigger was introduced.

https://securitylab.github.com/research/github-actions-preventing-pwn-requests/

**Review by luispadron — APPROVED** [review]
Thanks! Does this only support the case where the macro is in an external dependency or will it generate macro target for local packages with gazelle?

**Comment by jpsim** [comment]
I’ve only tested this with external dependencies.

**Comment by jpsim** [comment]
Split off just the functionality change into its own PR: https://github.com/cgrindel/rules_swift_package_manager/pull/753

Once CI supports Xcode 15 we can update the TCA example.

**Comment by jpsim** [comment]
#747 & #755 will complete this
